src/dune_pkg/opam_repo.ml

```diff
@@ -259,6 +268,26 @@ let load_all_versions ts opam_package_name =
   all_packages_versions_map ts opam_package_name |> load_all_versions_by_keys
 ;;
 
+let packages_in_repo repo =
+  let path = Paths.packages in
+  let packages =
+    match repo.source with
+    | Repo rev ->
+      all_packages_versions_at_rev_at_path ~path rev
+      |> List.map ~f:(fun (_opam_file, pkg) -> OpamPackage.name pkg)
+      |> OpamPackage.Name.Set.of_list
+      (* Set.to_list is only available in OCaml >= 5.1 *)
+      |> OpamPackage.Name.Set.to_list_map (fun x -> x)
```

Comment:
You are welcome to add a copy-paste implementation in stdune/set.ml. We do this for various other things like `string` and `Map`.